### ISSUE TITLE
hel-fns--autoloaded-p bugfix in Emacs29.0.50 appears to no longer be needed. 

### ISSUE THREAD

**Issue by mayagerard0** [hlissner/doom-emacs#6127]
### What did you expect to happen?

When I run the helpful-function (SPC h f)  it should open the help buffer with the documentation.

### What actually happened?

When I run (helpful-function 'doom/escape) , it fails and shows an error:

```
Debugger entered--Lisp error: (wrong-number-of-arguments #<subr help-fns--autoloaded-p> 1)
  #<subr help-fns--autoloaded-p>(doom/escape)
  funcall(#<subr help-fns--autoloaded-p> doom/escape)
  help-fns--autoloaded-p(doom/escape "/Users/gerardomaya/.emacs.d/core/core-keybinds.el")
  #f(compiled-function (sym buf) "Return non-nil if function SYM is autoloaded." #<bytecode 0x107b8cfcbffee66c>)(doom/escape #<buffer core-keybinds.el>)
  apply(#f(compiled-function (sym buf) "Return non-nil if function SYM is autoloaded." #<bytecode 0x107b8cfcbffee66c>) (doom/escape #<buffer core-keybinds.el>))
  (progn (fset #'help-fns--autoloaded-p vnew) (apply fn args))
  (unwind-protect (progn (fset #'help-fns--autoloaded-p vnew) (apply fn args)) (fset #'help-fns--autoloaded-p old))
  (let* ((vnew #'(lambda (sym _) (funcall help-fns--autoloaded-p sym))) (old (symbol-function #'help-fns--autoloaded-p))) (unwind-protect (progn (fset #'help-fns--autoloaded-p vnew) (apply fn args)) (fset #'help-fns--autoloaded-p old)))
  (let ((help-fns--autoloaded-p (symbol-function #'help-fns--autoloaded-p))) (ignore help-fns--autoloaded-p) (let* ((vnew #'(lambda (sym _) (funcall help-fns--autoloaded-p sym))) (old (symbol-function #'help-fns--autoloaded-p))) (unwind-protect (progn (fset #'help-fns--autoloaded-p vnew) (apply fn args)) (fset #'help-fns--autoloaded-p old))))
  doom--fix-helpful--autoloaded-p(#f(compiled-function (sym buf) "Return non-nil if function SYM is autoloaded." #<bytecode 0x107b8cfcbffee66c>) doom/escape #<buffer core-keybinds.el>)
  apply(doom--fix-helpful--autoloaded-p #f(compiled-function (sym buf) "Return non-nil if function SYM is autoloaded." #<bytecode 0x107b8cfcbffee66c>) (doom/escape #<buffer core-keybinds.el>))
  helpful--autoloaded-p(doom/escape #<buffer core-keybinds.el>)
  helpful--summary(doom/escape t #<buffer core-keybinds.el> 3115)
  #f(compiled-function () (interactive nil) #<bytecode 0xa0443cffb54862d>)()
  apply(#f(compiled-function () (interactive nil) #<bytecode 0xa0443cffb54862d>) nil)
  helpful-update()
  helpful--update-and-switch-buffer(doom/escape t)
  helpful-callable(doom/escape)
  funcall-interactively(helpful-callable doom/escape)
  command-execute(helpful-callable)
```

### Describe your attempts to resolve the issue

The error went away when I commented out the bug fix in core-editor.el: 

```
  (when EMACS29+
    (defadvice! doom--fix-helpful--autoloaded-p (fn &rest args)
      :around #'helpful--autoloaded-p
      (letf! (defun help-fns--autoloaded-p (sym _)
               (funcall help-fns--autoloaded-p sym))
        (apply fn args))))
```

### Steps to reproduce

1. Open Emacs29.0.50
2. Run SPC h f 
3. select any function
4. Error.

### System Information

https://pastebin.com/FZuQWGDJ

**Comment by hlissner** [comment]
I cannot reproduce this, and removing the advice brings back the original issue (#6097). Since I don't see any change in [help-fns--autoloaded-p upstream](https://github.com/emacs-mirror/emacs/blob/master/lisp/help-fns.el#L691-L693), I suspect this is a local and/or unrelated issue.

**Comment by hlissner** [comment]
> I guess the issue is that 29.0.50 is not an actual version but basically just the master branch if I understand correctly.

Ah, I see what you mean. Then give this a try:

```elisp
;; add to $DOOMDIR/config.el
(defadvice! fix-doom--fix-helpful--autoloaded-p (fn &rest args)
  :override #'doom--fix-helpful--autoloaded-p
  (letf! (defun help-fns--autoloaded-p (sym file)
           (condition-case _
               (funcall help-fns--autoloaded-p sym file)
             (wrong-number-of-arguments
              (funcall help-fns--autoloaded-p sym))))
    (apply fn args)))
```

If it works, I'll upstream it.

That said, I recommend sticking to 28 (iirc there's a `emacs28-git` on the AUR). I'm [phasing out support for 29](https://github.com/hlissner/doom-emacs/commit/16fe83ef0e4a4f543bf2ca9204c96d0e67e314e8) until 28 is released, because I've seen a new breaking change almost every week this year.

**Comment by Lythenas** [comment]
Unfortunately I can't test it anymore because I updated it :/ Maybe @mayagerard0 can?

Originally I just used the `emacs-git` package in the AUR as an easy way to get native-comp but I can use `emacs28-git` for that. I haven't kept up with emacs releases and I didn't understand the release numbering at first.

I think it's a good idea to not support emacs 29 at the moment.

**Comment by nyambol** [comment]
I can attest that this error occurs in vanilla GNU Emacs. 

GNU Emacs 29.0.50 (build 2, x86_64-w64-mingw32) of 2022-02-10

Unfortunately, reverting to v28 would mean rebuilding my package base, as other, recently updated, packages are not backward compatible as compiled by v29.  :(

I wonder if this has anything to do with the recent update to `elisp-refs`. (20220128.1736)

Thanks for a great package. 👍 I'll miss it. And I'll be keeping an eye out for its return.

**Comment by hlissner** [comment]
Did my snippet work for either of you? I'll still upload that if it works, I just can't promise the same for future 29 issues.

**Comment by mayagerard0** [comment]
The snippet works in my case.

**Comment by thaenalpha** [comment]
It seems like the `helpful` package is updated on this recently with the same
approach as in your snippet and the fix appears to no longer be needed again :rolling_on_the_floor_laughing:.
> 67cdd10 master origin/master Fix `helpful--autoloaded-p` in Emacs 29.0.50 (#283)

**Comment by hlissner** [comment]
As if b57d4c8 this issue should be resolved. Let me know if that isn't the case and I'll reopen it. Thanks for bringing it to my attention!
